File: medium-and-easy/003-longest-substirng-without-repeating-characters/Solution3.java
import java.util.HashMap;
import java.util.Map;

class Solution3 {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<Character, Integer>();

        int j = 0,
        answer = 0;

        for(int i = 0; i < s.length(); i++) {

            while(j < s.length() && !map.containsKey(s.charAt(j))) {
                map.put(s.charAt(j), j);
                answer = Math.max(answer, j-i+1);
                j++;
            }

            map.remove(s.charAt(i));

        }

        return answer;
    }
}

File: medium-and-easy/003-longest-substirng-without-repeating-characters/Solution3.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var map = new Map();

    var j = 0,
        answer = 0;

    for(var i = 0; i < s.length; i++) {

        while(j < s.length && !map.has(s.charAt(j))) {
            map.set(s.charAt(j), j);
            answer = Math.max(answer , j-i+1);
            ++j;
        }

        map.delete(s.charAt(i));

    }

    return answer;

};
